In [4]:
import numpy as np
from exputils.st_norm import compute_st_norm
from exputils.RoM.fwht import calculate_RoM_FWHT
from exputils.state.random import make_random_quantum_state
from tqdm.auto import tqdm

n_qubit = 8
SEED_CNT = 100

RoMs = []
st_norms = []
ratios = []
ratios_per_qubit = []

for seed in tqdm(range(SEED_CNT)):
    rho_vec = make_random_quantum_state("mixed", n_qubit, seed)
    st_norm = compute_st_norm(rho_vec)
    assert np.isclose(st_norm, np.linalg.norm(rho_vec, ord=1) / (2**n_qubit))
    R_FWHT = calculate_RoM_FWHT(n_qubit, rho_vec)[0]
    ratios.append(R_FWHT / (2 ** (n_qubit / 2) * st_norm))

print(np.min(ratios))
print(np.max(ratios))

  0%|          | 0/100 [00:00<?, ?it/s]

0.9943682073176954
1.0014677091887252


for ten qubits random mixed quantum state,

$$
    0.994 \leq 
    \frac{\mathcal{R}(\rho)}{2^{n/2}||\rho||_{\text{st}}}
    \leq 1.002
$$
